pkg/upgrade/control_plane.go

```diff
@@ -279,67 +330,113 @@ func (u *ControlPlaneUpgrader) updateMachines(machines *clusterapiv1alpha1.Machi
-		if err != nil {
-			return err
-		}
-
-		oldNode := u.GetNodeFromProviderID(originalProviderID.ID())
-		if oldNode == nil {
-			u.log.Info("Couldn't retrieve oldNode", "id", originalProviderID.String())
-			return fmt.Errorf("unknown previous node %q", originalProviderID.String())
+		// assume the original name is controlplane-<index> or controlplane-<index>-<timestamp>
+		// let's set the new name to controlplane-<index>-<timestamp>
+		nameParts := strings.Split(machine.Name, "-")
+		if len(nameParts) < 2 {
+			return errors.Errorf("machine name %q does not match expected format <name>-<index> or <name>-<index>-<timestamp>", machine.Name)
 		}
+		name := fmt.Sprintf("%s-%s-%d", nameParts[0], nameParts[1], time.Now().Unix())
+		// TODO: generate the name based off each respective object
 
-		oldHostName := hostnameForNode(oldNode)
-
-		newMachine, node, err := machineCreator.NewMachine(&machine)
+		err = u.updateInfrastructureReference(name, &machine)
 		if err != nil {
 			return err
 		}
-		nodeHostname := hostnameForNode(node)
-
-		// This used to happen when a new machine was created as a side effect. Must still update the mapping.
-		if err := u.UpdateProviderIDsToNodes(); err != nil {
-			return err
-		}
 
-		// delete old etcd member
-		err = u.deleteEtcdMember(time.Minute*1, nodeHostname, u.oldNodeToEtcdMember[oldHostName])
+		err = u.updateBootstrapConfig(name, &machine)
 		if err != nil {
-			return errors.Wrapf(err, "unable to delete old etcd member %s", u.oldNodeToEtcdMember[oldHostName])
-		}
-
-		if err := u.deleteMachine(&machine); err != nil {
 			return err
 		}
 
-		if err := u.applyAnnotation(newMachine); err != nil {
+		if err := u.updateMachine(name, machine, machineCreator); err != nil {
 			return err
 		}
 	}
+
 	return nil
 }
 
-func (u *ControlPlaneUpgrader) applyAnnotation(m *clusterapiv1alpha1.Machine) error {
-	original, err := json.Marshal(m)
+func (u *ControlPlaneUpgrader) updateBootstrapConfig(name string, machine *clusterapiv1alpha2.Machine) error {
+	// copy node registration
+	oldBootstrap := &bootstrapv1.KubeadmConfig{}
+	if machine.Spec.Bootstrap.ConfigRef.Namespace == "" {
+		machine.Spec.Bootstrap.ConfigRef.Namespace = "default"
+	}
+	err := u.ctrlClient.Get(context.TODO(), ctrlclient.ObjectKey{Name: machine.Spec.Bootstrap.ConfigRef.Name, Namespace: machine.Spec.Bootstrap.ConfigRef.Namespace}, oldBootstrap)
 	if err != nil {
-		return errors.WithStack(err)
+		return err
 	}
-	if m.Annotations == nil {
-		m.Annotations = map[string]string{}
+
+	// create init configuration
+	newBootstrap := &bootstrapv1.KubeadmConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: oldBootstrap.Namespace,
+		},
+		Spec: bootstrapv1.KubeadmConfigSpec{
+			JoinConfiguration: &kubeadmv1beta1.JoinConfiguration{
+				ControlPlane:     &kubeadmv1beta1.JoinControlPlane{},
+				NodeRegistration: oldBootstrap.Spec.InitConfiguration.NodeRegistration,
```

Comment:
I think there should be some nil checks for Init/Join configs here on `oldBootstrap`. This will work for the first controlplane machine (which will have an Init config for sure), but it won't work for subsequent controlplane machines because they have JoinConfigurations.

Reply:
oh no, my above assessment is totally incorrect. these are all JoinConfigurations.

I think there's a chance that InitConfiguration.NodeRegistration may be nil though if the original KubeadmConfig only has a JoinConfiguration.